From 6d343d123f1d7e0dc42d8da073c06bc5074127b5 Mon Sep 17 00:00:00 2001
From: Li Wencheng <liwencheng@phytium.com.cn>
Date: Fri, 26 Apr 2024 15:02:13 +0800
Subject: [PATCH 18/67] PHYTIUM: net/phytmac: Support usxgmii wol feature

Support usxgmii wol by change the value of registers,
including  wol register, int register, net config register
and spec_add1 register.

Mainline: NA
Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Change-Id: Ie651779a4bc2487aaaa51b24c69a5cc9d8de3784

Link: https://gitee.com/phytium_embedded/phytium-linux-kernel/commit/11ba74ddc3fcd568a41284b08ff69a86809e38fa
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 drivers/net/ethernet/phytium/phytmac_main.c | 12 ++++++------
 drivers/net/ethernet/phytium/phytmac_v1.c   |  9 +++++++++
 drivers/net/ethernet/phytium/phytmac_v1.h   | 11 +++++++++++
 3 files changed, 26 insertions(+), 6 deletions(-)

diff --git a/drivers/net/ethernet/phytium/phytmac_main.c b/drivers/net/ethernet/phytium/phytmac_main.c
index 058f9647b855..28b6dc86f668 100644
--- a/drivers/net/ethernet/phytium/phytmac_main.c
+++ b/drivers/net/ethernet/phytium/phytmac_main.c
@@ -1732,12 +1732,6 @@ static int phytmac_open(struct net_device *ndev)

 	hw_if->reset_hw(pdata);

-	ret = phytmac_get_mac_address(pdata);
-	if (ret) {
-		netdev_err(ndev, "phytmac get mac address failed\n");
-		goto reset_hw;
-	}
-
 	ret = netif_set_real_num_tx_queues(ndev, pdata->queues_num);
 	if (ret) {
 		netdev_err(ndev, "error setting real tx queue number\n");
@@ -2123,6 +2117,12 @@ int phytmac_drv_probe(struct phytmac *pdata)
 		goto err_phylink_init;
 	}

+	ret = phytmac_get_mac_address(pdata);
+	if (ret) {
+		netdev_err(ndev, "phytmac get mac address failed\n");
+		goto err_phylink_init;
+	}
+
 	if (netif_msg_probe(pdata))
 		dev_dbg(pdata->dev, "probe successfully! Phytium %s at 0x%08lx irq %d (%pM)\n",
 			"MAC", ndev->base_addr, ndev->irq, ndev->dev_addr);
diff --git a/drivers/net/ethernet/phytium/phytmac_v1.c b/drivers/net/ethernet/phytium/phytmac_v1.c
index 4a690b6a96cf..b96547e54d82 100644
--- a/drivers/net/ethernet/phytium/phytmac_v1.c
+++ b/drivers/net/ethernet/phytium/phytmac_v1.c
@@ -489,6 +489,15 @@ static int phytmac_set_wake(struct phytmac *pdata, int wake)
 		value |= PHYTMAC_BIT(MCAST);

 	PHYTMAC_WRITE(pdata, PHYTMAC_WOL, value);
+	if (wake) {
+		PHYTMAC_WRITE(pdata, PHYTMAC_IE, PHYTMAC_BIT(WOL_RECEIVE_ENABLE));
+		value = PHYTMAC_READ(pdata, PHYTMAC_NCONFIG) | PHYTMAC_BIT(IGNORE_RX_FCS);
+		PHYTMAC_WRITE(pdata, PHYTMAC_NCONFIG, value);
+	} else {
+		PHYTMAC_WRITE(pdata, PHYTMAC_ID, PHYTMAC_BIT(WOL_RECEIVE_DISABLE));
+		value = PHYTMAC_READ(pdata, PHYTMAC_NCONFIG) & ~PHYTMAC_BIT(IGNORE_RX_FCS);
+		PHYTMAC_WRITE(pdata, PHYTMAC_NCONFIG, value);
+	}

 	return 0;
 }
diff --git a/drivers/net/ethernet/phytium/phytmac_v1.h b/drivers/net/ethernet/phytium/phytmac_v1.h
index 2cabadfed2f1..1f49d4ec79a0 100644
--- a/drivers/net/ethernet/phytium/phytmac_v1.h
+++ b/drivers/net/ethernet/phytium/phytmac_v1.h
@@ -140,6 +140,8 @@ extern struct phytmac_hw_if phytmac_1p0_hw;
 #define PHYTMAC_DBW_128				4
 #define PHYTMAC_RCO_EN_INDEX			24	 /* Receive checksum offload enable */
 #define PHYTMAC_RCO_EN_WIDTH			1
+#define PHYTMAC_IGNORE_RX_FCS_INDEX     26
+#define PHYTMAC_IGNORE_RX_FCS_WIDTH     1
 #define PHYTMAC_SGMII_EN_INDEX			27	 /* Sgmii mode enable */
 #define PHYTMAC_SGMII_EN_WIDTH			1

@@ -364,6 +366,15 @@ extern struct phytmac_hw_if phytmac_1p0_hw;
 #define PHYTMAC_TXTAIL_ENABLE_INDEX		0 /* Enable tx tail */
 #define PHYTMAC_TXTAIL_ENABLE_WIDTH		1

+/* Bitfields in INT ENABLE */
+#define PHYTMAC_WOL_RECEIVE_ENABLE_INDEX    28 /* Enable wol_event_recieve */
+#define PHYTMAC_WOL_RECEIVE_ENABLE_WIDTH    1
+
+/* Bitfields in INT DISABLE */
+#define PHYTMAC_WOL_RECEIVE_DISABLE_INDEX    28 /* Disable wol_event_recieve */
+#define PHYTMAC_WOL_RECEIVE_DISABLE_WIDTH    1
+
+
 #define PHYTMAC_TSEC_WIDTH (PHYTMAC_SECH_WIDTH + PHYTMAC_SECL_WIDTH)
 #define SEC_MAX_VAL (((u64)1 << PHYTMAC_TSEC_WIDTH) - 1)
 #define NSEC_MAX_VAL ((1 << PHYTMAC_NSEC_WIDTH) - 1)
--
2.50.1
